iroh-io/src/mem.rs

```diff
@@ -0,0 +1,102 @@
+use super::{AsyncSliceReader, AsyncSliceWriter};
+use bytes::{Bytes, BytesMut};
+use futures::future;
+use std::io;
+
+impl AsyncSliceReader for bytes::Bytes {
```

Comment:
Do you plan on adding impls for `Vec<u8>` and `std::io::Cursor` potentially? might be quite nice

Reply:
Yeah, for Vec it would basically be the same. Cursor I am not sure about since you don't have the notion of a "current position".

Reply:
it would also be good to test that this interface works with https://docs.rs/mmap-rs (and potentially adding a cfged impl to this code) as this would likely be a nice speedup in a bunch of cases.

Reply:
just a thought though

Reply:
Oh boy, that is a can of worms.

So - it is definitely possible to do an impl of this that uses memory mapping. There is a big *but* though: since there is no custom constructor for `Bytes` you would have to copy from the memory mapped region into the Bytes, even if you could ensure that the Bytes will keep the memory mapped buffer alive.

I have opened an issue on bytes::Bytes for this, and so have other people. But so far no success. For other projects I have therefore decided to write my own Bytes equivalent type. But for this crate I decided to go with Bytes since it is ubiquitous and the Bytes/BytesMut combo is quite helpful when doing zero alloc IO (splitting off Bytes from a BytesMut etc) and I did not feel like reimplementing all of this.

This was my issue:
https://github.com/tokio-rs/bytes/issues/571

The other issues:
https://github.com/tokio-rs/bytes/issues/437
https://github.com/tokio-rs/bytes/issues/526

My version of Bytes:
https://github.com/cloudpeers/radixdb/blob/master/src/store/blob_store.rs#L37

Reply:
By the way: what is the current crate for memory mapped IO?

There is memmap, memmap2, and now this mmap-rs

I thought memmap2 was the current one... 🤷‍♂️

Reply:
This is memmap2 download stats. It certainly seems to be popular...

![image](https://github.com/n0-computer/iroh/assets/248257/91f6ed72-7f15-4ccb-9a44-0984a7318cd1)